gctree/branching_processes.py

```diff
@@ -920,42 +962,466 @@ def ll(
         """
-        if self.forest is None:
-            raise ValueError("forest data must be defined to compute likelihood")
-        CollapsedTree._build_ll_genotype_cache(self._c_max, self._m_max, p, q)
-        # we don't want to build the cache again in each tree
-        terms = [tree.ll(p, q, build_cache=False) for tree in self.forest]
-        ls = np.array([term[0] for term in terms])
-        grad_ls = np.array([term[1] for term in terms])
+        if self._cm_countlist is None:
+            if self._forest is not None:
+                cmcount_dagfuncs = _cmcounter_dagfuncs()
+
+                def to_tuple(mset):
+                    # When there's unobserved root unifurcation, augment with
+                    # pseudocount.
+                    if (0, 1) in mset:
+                        assert mset[(0, 1)] == 1
+                        mset = mset - {(0, 1)} + {(1, 1)}
+                    return tuple(mset.items())
+
+                cmcounters = self._forest.weight_count(**cmcount_dagfuncs)
+                self._cm_countlist = [
+                    (to_tuple(mset), mult) for mset, mult in cmcounters.items()
+                ]
+                # an iterable containing tuples `(cm_counts, mult)`, where `cm_counts`
+                # is an iterable describing a tree, and `mult` is the number of trees in the forest
+                # which can be described with the iterable `cm_counts`.
+            elif self._ctrees is not None:
+                self._cm_countlist = tuple(
+                    coll.Counter([tree._cm_counts for tree in self._ctrees]).items()
+                )
+            else:
+                raise ValueError("forest data must be defined to compute likelihood")
+
+        terms = [
+            [_lltree(cmcounts, p, q), count] for cmcounts, count in self._cm_countlist
+        ]
+        ls = np.array([term[0][0] for term in terms])
+        grad_ls = np.array([term[0][1] for term in terms])
+        # This can be done ahead of time
+        count_ls = np.array([term[1] for term in terms])
         if marginal:
             # we need to find the smallest derivative component for each
             # coordinate, then subtract off to get positive things to logsumexp
             grad_l = []
             for j in range(len((p, q))):
                 i_prime = grad_ls[:, j].argmin()
+                b = (grad_ls[:, j] - grad_ls[i_prime, j]) * count_ls
+                # believe it or not, logsumexp can't handle 0 in b
+                # when np.seterr(underflow='raise') on newer processors:
                 grad_l.append(
                     grad_ls[i_prime, j]
                     + np.exp(
-                        logsumexp(
-                            ls - ls[i_prime], b=grad_ls[:, j] - grad_ls[i_prime, j]
-                        )
-                        - logsumexp(ls - ls[i_prime])
+                        scs.logsumexp((ls - ls[i_prime])[b != 0], b=b[b != 0])
```

Comment:
This line is causing the error. When b consists entirely of zeros, then we're passing empty lists to logsumexp.  Not sure in what situation it's all zeros yet

Reply:
fixed by `374cc06`. When the coefficient list `b` is all zeros, I was passing empty lists to logsumexp. Now I test for that situation and replace the offending calls with a hard-coded value